include/envoy/http/async_client.h

```diff
@@ -131,34 +131,83 @@ class AsyncClient {
 
   virtual ~AsyncClient() {}
 
+  /**
+   * A structure to hold the options for AsyncStream object.
+   */
+  struct StreamOptions {
+    StreamOptions() {}
+    StreamOptions(const absl::optional<std::chrono::milliseconds>& timeout) : timeout(timeout) {}
+    StreamOptions(const absl::optional<std::chrono::milliseconds>& timeout, bool buffer_body)
+        : timeout(timeout), buffer_body_for_retry(buffer_body) {}
+    StreamOptions(const absl::optional<std::chrono::milliseconds>& timeout, bool buffer_body,
+                  bool send_xff)
+        : timeout(timeout), buffer_body_for_retry(buffer_body), send_xff(send_xff) {}
+    StreamOptions(bool buffer_body) : buffer_body_for_retry(buffer_body) {}
+    StreamOptions(bool buffer_body, bool send_xff)
+        : buffer_body_for_retry(buffer_body), send_xff(send_xff) {}
+
+    // For gmock test
+    bool operator==(const StreamOptions& src) const {
+      return timeout == src.timeout && buffer_body_for_retry == src.buffer_body_for_retry &&
+             send_xff == src.send_xff;
+    }
+
+    // The timeout supplies the stream timeout, measured since when the frame with
+    // end_stream flag is sent until when the first frame is received.
+    absl::optional<std::chrono::milliseconds> timeout;
+
+    // The buffer_body_for_retry specifies whether the streamed body will be buffered so that
+    // it can be retried. In general, this should be set to false for a true stream. However,
+    // streaming is also used in certain cases such as gRPC unary calls, where retry can
+    // still be useful.
+    bool buffer_body_for_retry{false};
+
+    // If true, x-forwarded-for header will be added.
+    bool send_xff{true};
+  };
+
+  /**
+   * A structure to hold the options for AsyncRequest object.
+   */
+  struct RequestOptions : public StreamOptions {
+    RequestOptions() {}
+    RequestOptions(const absl::optional<std::chrono::milliseconds>& timeout)
+        : StreamOptions(timeout) {}
+    RequestOptions(const absl::optional<std::chrono::milliseconds>& timeout, bool buffer_body)
+        : StreamOptions(timeout, buffer_body) {}
+    RequestOptions(const absl::optional<std::chrono::milliseconds>& timeout, bool buffer_body,
+                   bool send_xff)
+        : StreamOptions(timeout, buffer_body, send_xff) {}
+    RequestOptions(bool buffer_body, bool send_xff) : StreamOptions(buffer_body, send_xff) {}
+
+    // For gmock test
+    bool operator==(const RequestOptions& src) const {
+      return *dynamic_cast<const StreamOptions*>(this) == *dynamic_cast<const StreamOptions*>(&src);
```

Comment:
this is a static_cast (upcast)? You can just do `return StreamOptions::operator==(src)` as well.

Reply:
Done